### PR TITLE
fedekunze/show rewards

### LINKED ISSUE #135: Show commission earned
I as a user
Want to see the amount of commission I earned
So that I stay engaged in holding atoms.

AC:
- [ ] a user should be able to see how many rewards they have been issued from each validator they have delegated to
- [ ] a user should be able to see a sum of the total rewards he has in all validators
- [ ] rewards should update in the UI as they update

See following event for updates on rewards (taken from https://github.com/cosmos/cosmos-sdk/blob/develop/docs/spec/distribution/state.md):

dependencies: #209

### PR DESCRIPTION
**Pull request by fedekunze**
Closes #135

**Description:**

Show delegation rewards on validators pages and outstanding rewards on network page

screen shot 2019-02-28 at 3 57 47 pm
screen shot 2019-02-28 at 3 57 54 pm
screen shot 2019-02-28 at 3 58 01 pm

Thank you! 🚀

---

For contributor:

- [x] Added entries in `CHANGELOG.md` with issue # and GitHub username
- [x] Reviewed `Files changed` in the github PR explorer
- [x] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

File: app/src/renderer/components/network/PageNetwork.vue [hunk 1]
```diff
@@ -132,7 +132,10 @@ export default {
         }
       ]
     }
-  }
+  },
+  async mounted() {
```
Comment (jbibla) [inline app/src/renderer/components/network/PageNetwork.vue]: nice!

this will only get the data once... right? should we update it every block / every 10 blocks? is this too many requests?

File: app/src/renderer/components/staking/LiValidator.vue [hunk 2]
```diff
@@ -43,7 +43,7 @@
       }}
     </td>
     <td class="li-validator__rewards data-table__row__cell__separator">
-      n/a
+      {{ rewards }}
```
Comment (jbibla) [inline app/src/renderer/components/staking/LiValidator.vue]: 🎉

File: app/src/renderer/components/staking/LiValidator.vue [hunk 3]
```diff
@@ -97,6 +97,13 @@ export default {
       }
       return `n/a`
     },
+    rewards() {
+      const validatorRewards = this.distribution.rewards[this.validator.operator_address]
+      if(validatorRewards) {
+        return this.num.prettyInt(validatorRewards[`stake`])
```
Comment (jbibla) [inline app/src/renderer/components/staking/LiValidator.vue]: should this use bondingDenom instead of `stake`? or some other param?

File: app/src/renderer/components/staking/LiValidator.vue [hunk 4]
```diff
@@ -97,6 +97,13 @@ export default {
...
+      }
+      return `n/a`
```
Comment (jbibla) [inline app/src/renderer/components/staking/LiValidator.vue]:
```suggestion
      return `--`
```

File: app/src/renderer/scripts/common.js [hunk 5]
```diff
@@ -27,18 +37,6 @@ module.exports.ratToBigNumber = function(rat) {
   }
   return new BN(rat)
 }
-// TODO uncomment when validator comission is done
-// module.exports.parseValidatorShares = function(validator) {
```
Comment (jbibla) [inline app/src/renderer/scripts/common.js]: are these functions not usable now?

File: app/src/renderer/vuex/modules/distribution.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
+import * as Sentry from "@sentry/browser"
+import Vue from "vue"
+import { coinsToObject } from "scripts/common.js"
+
+export default ({ node }) => {
+    const emptyState = {
+        loading: false,
+        loaded: false,
+        error: null,
+        /* totalRewards use the following format:
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: i don't know if i like having these comments like this.

File: app/src/renderer/components/staking/TabMyDelegations.vue [hunk 7]
```diff
@@ -99,6 +99,12 @@ export default {
               transaction.tx.value.msg[0].value.validator_addr
             ]
         }))
+  },
+  async mounted() {
+    this.$store.dispatch(`getTotalRewards`)
+    this.yourValidators.forEach(validator => {
+      this.$store.dispatch(`getRewardsFromValidator`, validator.operator_address)
```
Comment (jbibla) [inline app/src/renderer/components/staking/TabMyDelegations.vue]: this is good for now. might be an issue if a user has too many validators... maybe a good candidate for a new special endpoint on the REST API. 

@alexanderbez 

maybe the endpoint `/distribution/delegators/cosmos1address/rewards` could return the data per validator instead of as a single total value? just an idea.

Comment (jbibla) [inline app/src/renderer/components/staking/TabMyDelegations.vue]: not sure about making multiple requests for this. should we request a new endpoint for this?

File: app/src/renderer/vuex/modules/distribution.js [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            { state, rootState: { session }, commit }
+        ) {
+            state.loading = true
+            try {
+                const rewardsArray = await node.getDelegatorRewards(session.address)
+                const rewards = coinsToObject(rewardsArray)
+                commit(`setTotalRewards`, rewards)
+                state.error = null
+            } catch (error) {
+                commit(`notifyError`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: no more notify

File: app/src/renderer/vuex/modules/distribution.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        },
+        async getRewardsFromValidator({ state, rootState: { session }, commit }, validatorAddr) {
+            state.loading = true
+            try {
+                const rewardsArray = await node.getDelegatorRewardsFromValidator(session.address, validatorAddr)
+                const rewards = coinsToObject(rewardsArray)
+                commit(`setDelegationRewards`, { validatorAddr, rewards })
+                state.error = null
+            } catch (error) {
+                commit(`notifyError`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: no more notify

File: app/src/renderer/vuex/modules/distribution.js [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                state.error = error
+            }
+            state.loading = false
+            state.loaded = true
+        },
+        async withdrawRewardsFromValidator(
+            { rootState: { wallet }, commit, dispatch },
+            { validatorAddr, password, submitType }
+        ) {
+            await dispatch(`sendTx`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: not for this issue - but this seems kind of dangerous from an implementation perspective. can we make `sendTx` a more generalized function that takes `type` and other relevant `...args`? having to add `wallet.address` and `password` like this is maybe problematic?

@faboweb @sabau ?

File: app/src/renderer/vuex/modules/distribution.js [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                type: `postWithdrawDelegatorRewardsFromValidator`,
+                to: wallet.address,
+                pathParameter: validatorAddr,
+                password,
+                submitType
+            })
+            // optimistic update
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: 😄

File: app/src/renderer/vuex/modules/distribution.js [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+            commit(`setDelegationRewards`, { validatorAddr, rewards: {} })
+            // update rewards
+            await dispatch(`getTotalRewards`)
+            await dispatch(`getRewardsFromValidator`, validatorAddr)
+        },
+        // TODO: move to a common parameters module
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: what is your thinking here? what params are there?

File: app/src/renderer/vuex/modules/distribution.js [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        async getDistributionParameters({ commit }) {
+            state.loading = true
+            try {
+                const parameters = await node.getDistributionParameters()
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: **not for this PR** - i would love to rename `node` to something a lot more descriptive. 

@faboweb @sabau @fedekunze any ideas?

File: app/src/renderer/vuex/modules/distribution.js [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+                commit(`setDistributionParameters`, parameters)
+                state.error = null
+            } catch (error) {
+                commit(`notifyError`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: no mo' notify

Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: use https://github.com/cosmos/voyager/pull/2049/files#diff-b37c59a991d76e68f3b9a84431a39ab5R9 instead

File: app/src/renderer/vuex/modules/distribution.js (hunk 8 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: use https://github.com/cosmos/voyager/pull/2049/files#diff-b37c59a991d76e68f3b9a84431a39ab5R9 instead

File: app/src/renderer/vuex/modules/distribution.js (hunk 9 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: use https://github.com/cosmos/voyager/pull/2049/files#diff-b37c59a991d76e68f3b9a84431a39ab5R9 instead

File: app/src/renderer/vuex/modules/distribution.js [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            state.loaded = true
+        },
+        async getOutstandingRewards({ commit }) {
+            state.loading = true
+            try {
+                const oustandingRewardsArray = await node.getDistributionOutstandingRewards()
+                const oustandingRewards = coinsToObject(oustandingRewardsArray)
+                commit(`setOutstandingRewards`, oustandingRewards)
+            } catch (error) {
+                commit(`notifyError`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/distribution.js]: no mo` notify 

use https://github.com/cosmos/voyager/pull/2049/files#diff-b37c59a991d76e68f3b9a84431a39ab5R9 instead

File: app/src/renderer/vuex/modules/send.js [hunk 16]
```diff
@@ -89,13 +89,23 @@ export default ({ node }) => {
       const submitType = args.submitType || `local`
       delete args.submitType
 
-      // extract "to" address
+      // TODO: refactor to get path request parameters at once
+      // extract path parameters
       const to = args.to
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/send.js]: would `toAddress` be more descriptive?

File: app/src/renderer/vuex/modules/validators.js [hunk 17]
```diff
@@ -54,6 +59,29 @@ export default ({ node }) => {
       if (validatorHash === state.validatorHash) return
       commit(`setValidatorHash`, validatorHash)
       await dispatch(`getValidators`)
+    },
+    async getValidatorDistributionInfoAndRewards({ commit }, validatorAddr) {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/validators.js]: can this be split into two functions? 

`getValidatorDistributionInfo()`
`getValidatorRewards()`

File: app/src/renderer/vuex/modules/validators.js [hunk 18]
```diff
@@ -54,6 +59,29 @@ export default ({ node }) => {
...
+        const rewardsArray = await node.getValidatorRewards(validatorAddr)
+
+        const rewards = coinsToObject(rewardsArray)
+        self_bond_rewards = coinsToObject(self_bond_rewards)
+        val_commission = coinsToObject(val_commission)
+
+        commit(`setValidatorDistributionInfo`, { self_bond_rewards, val_commission, rewards })
+        state.error = null
+      } catch (error) {
+        commit(`notifyError`, {
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/validators.js]: no mo' notify

https://github.com/cosmos/voyager/pull/2049/files#diff-b834fe9e3ec5cca6b2d0a8b6526c1208R9

**Comment by jbibla** [comment]
please add rewards to the balance header too!

File: app/src/renderer/vuex/modules/distribution.js (hunk 13 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: `api`?

Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: that's what i was thinking too

File: app/src/renderer/components/network/PageNetwork.vue (hunk 1 above)
Comment (faboweb) [reply app/src/renderer/components/network/PageNetwork.vue]: issue states: `* rewards should update in the UI as they update`

File: app/src/renderer/scripts/common.js (hunk 5 above)
Comment (faboweb) [reply app/src/renderer/scripts/common.js]: they are decimals now and not `1000/2500000`

File: app/src/renderer/vuex/modules/distribution.js (hunk 6 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: if we don't have types this is a good documentation on the formats. what is bugging you with them?

File: app/src/renderer/components/staking/TabMyDelegations.vue [hunk 19]
```diff
@@ -99,6 +99,12 @@ export default {
               transaction.tx.value.msg[0].value.validator_addr
             ]
         }))
+  },
+  async mounted() {
```
Comment (faboweb) [inline app/src/renderer/components/staking/TabMyDelegations.vue]: This is also just querying the rewards once.

File: app/src/renderer/scripts/common.js [hunk 20]
```diff
@@ -18,6 +18,16 @@ module.exports.shortAddress = function(address, length = 4) {
   }
 }
 
+module.exports.coinsToObject = function(coinArray) {
+  const coins = {}
```
Comment (faboweb) [inline app/src/renderer/scripts/common.js]: proposal for a reduce:
```
const coins = coinArray.reduce((dictionary,{ denom, amount }) => { dictionary[denom] = amount; return dictionary; }, {})
return coins
```

File: app/src/renderer/vuex/modules/distribution.js [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+        resetSessionData({ rootState }) {
+            rootState.delegation = JSON.parse(JSON.stringify(emptyState))
+        },
+        async getTotalRewards(
+            { state, rootState: { session }, commit }
+        ) {
+            state.loading = true
+            try {
+                const rewardsArray = await node.getDelegatorRewards(session.address)
+                const rewards = coinsToObject(rewardsArray)
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/distribution.js]: I like the `coinsToObject` refactor

**Comment by fedekunze** [comment]
thanks for your comments ! I will address the distr module comments on #2047

File: app/src/renderer/vuex/modules/distribution.js [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+            state.loading = false
+            state.loaded = true
+        },
+        async withdrawAllRewards(
+            { rootState: { wallet }, dispatch },
+            { password, submitType }
+        ) {
+            await dispatch(`sendTx`, {
+                type: `postWithdrawDelegatorRewards`,
+                to: wallet.address,
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/distribution.js]: please always use `session.address` so we don't store the address in 2 places.

File: app/src/renderer/vuex/modules/distribution.js [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+            state.loading = false
+            state.loaded = true
+        },
+        async withdrawRewardsFromValidator(
+            { rootState: { wallet }, commit, dispatch },
+            { validatorAddr, password, submitType }
+        ) {
+            await dispatch(`sendTx`, {
+                type: `postWithdrawDelegatorRewardsFromValidator`,
+                to: wallet.address,
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/distribution.js]:
```suggestion
                to: session.address,
```

File: app/src/renderer/vuex/modules/distribution.js (hunk 8 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/distribution.js]: Agree, but I'd also propose to leave notifications for `devMode` only. I like to see if something fails when I'm coding. Thoughts @faboweb @sabau ?

File: app/src/renderer/vuex/modules/distribution.js (hunk 10 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: we need to refactor this. but we will always need to send the password along like this if we don't store it.

Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: and `address` is a parameter of `postWithdrawDelegatorRewardsFromValidator`. but open for ideas.

File: app/src/renderer/vuex/modules/distribution.js (hunk 8 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/distribution.js]: I'll delete them here but maybe we can update all of them in a later refactor

File: app/src/renderer/vuex/modules/distribution.js [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+                    body: error.message
+                })
+                Sentry.captureException(error)
+                state.error = error
+            }
+            state.loading = false
+            state.loaded = true
+        },
+        async getOutstandingRewards({ commit }) {
+            state.loading = true
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/distribution.js]: note: we are using the same variable `loading` for rewards and parameters. so they are influencing each other. this might result into inconveniences.

File: app/src/renderer/vuex/modules/send.js [hunk 25]
```diff
@@ -89,13 +89,23 @@ export default ({ node }) => {
...
+      const pathParameter = args.pathParameter
       delete args.to
+      delete args.pathParameter
 
       // get the generated tx by querying it from the backend
-      const req = to ? node[type](to, args) : node[type](args)
-      const generationRes = await req.catch(handleSDKError)
+      let request
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/send.js]: this looks very complex and screams refactor :)

File: app/src/renderer/vuex/modules/validators.js (hunk 17 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/validators.js]: I agree

File: app/src/renderer/vuex/modules/distribution.js (hunk 22 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: can we remove `.address` from the wallet module?

File: app/src/renderer/vuex/modules/distribution.js (hunk 24 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: ah! multiple loading properties are required then!

File: app/src/renderer/vuex/modules/send.js (hunk 25 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/send.js]: 😱

**Comment by fedekunze** [comment]
ledger uses wallet.address too

File: app/src/renderer/vuex/modules/distribution.js (hunk 22 above)
Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: Yes please

File: app/src/renderer/vuex/modules/distribution.js (hunk 6 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: they take up many lines and we don't do this anywhere else.

File: app/src/renderer/vuex/modules/distribution.js (hunk 8 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: > I like to see if something fails when I'm coding.

we need to display the error message and corresponding component in the UI, just not as a notify.

File: app/src/renderer/vuex/modules/distribution.js (hunk 10 above)
Comment (jbibla) [reply app/src/renderer/vuex/modules/distribution.js]: i understand. i'm just proposing that we send a tx with type - but data that exists in getters can be populated elsewhere - instead of in the component.

Comment (faboweb) [reply app/src/renderer/vuex/modules/distribution.js]: > data that exists in getters can be populated elsewhere

sure

File: app/src/renderer/components/staking/TabMyDelegations.vue (hunk 7 above)
Comment (fedekunze) [reply app/src/renderer/components/staking/TabMyDelegations.vue]: imo we should eventually use something like graphQL

File: app/src/renderer/vuex/modules/distribution.js (hunk 12 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/distribution.js]: move all parameters into the parameters module and call all of them using for example `parameters.staking`, `parameters.distribution`, etc

File: app/src/renderer/vuex/modules/send.js (hunk 16 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: will address that on the refactor PR for this, I'd like to get the path params for the endpoints at once

File: app/src/renderer/vuex/modules/validators.js (hunk 17 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/validators.js]: done

File: app/src/renderer/vuex/modules/distribution.js (hunk 22 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/distribution.js]: #2114

File: app/src/renderer/components/network/PageNetwork.vue [hunk 26]
```diff
@@ -40,15 +40,23 @@
             <dl class="info_dl">
               <dt>Total Liquid {{ bondDenom }}</dt>
               <dd id="loose_tokens">
-                {{ pool.pool.loose_tokens ? pool.pool.loose_tokens : `n/a` }}
+                {{ pool.pool.not_bonded_tokens ? num.pretty(pool.pool.not_bonded_tokens) : `n/a` }}
```
Comment (sabau) [inline app/src/renderer/components/network/PageNetwork.vue]: Line 43 and 51 share the same logic (and 59 is partially doing that via `outstandingRewards`), can be moved to a function that takes as argument an integer

File: app/src/renderer/components/staking/LiValidator.vue [hunk 27]
```diff
@@ -98,10 +105,10 @@ export default {
...
     },
     yourVotes() {
-      return this.committedDelegations[this.validator.id]
+      return this.committedDelegations[this.validator.operator_address]
         ? calculateTokens(
           this.validator,
-          this.committedDelegations[this.validator.id]
+          this.committedDelegations[this.validator.operator_address]
         )
         : BigNumber(0)
```
Comment (sabau) [inline app/src/renderer/components/staking/LiValidator.vue]: Why BigNumber to represent a 0? is not part of this PR, just curious

File: test/unit/specs/store/distribution.spec.js [hunk 28]
```diff
@@ -106,6 +137,35 @@ describe(`Module: Fee Distribution`, () => {
...
+          validators[1].operator_address
+        )
+      })
+
+      it(`fails`, async () => {
+        const validators = [
+          { operator_address: undefined },
+          { operator_address: `cosmosvaloper1address2` },
+        ]
+        dispatch = jest.fn(async () => Promise.reject(Error(`invalid address`)))
```
Comment (sabau) [inline test/unit/specs/store/distribution.spec.js]: dispatch does not need the jest.fn wrapping, the simplified version should work
```suggestion
        dispatch = () => Promise.reject(Error(`invalid address`))
```

File: app/src/renderer/vuex/modules/distribution.js (hunk 6 above)
Comment (sabau) [reply app/src/renderer/vuex/modules/distribution.js]: I think here the problem we have is `we don't do this anywhere else`, it's true, but we should

File: app/src/renderer/components/common/TmBalance.vue [hunk 29]
```diff
@@ -16,7 +16,7 @@
       </div>
       <div v-if="rewards" class="top-section">
         <h3>Total Rewards</h3>
-        <h2>{{ rewards }}</h2>
+        <h2>{{ num.short(distribution.totalRewards[bondDenom] || 0) }}</h2>
```
Comment (faboweb) [inline app/src/renderer/components/common/TmBalance.vue]: we agreed on not using `short` and wait for #2044 
also I think putting this formatting in the computed as it was before is better for testability

File: app/src/renderer/vuex/modules/wallet.js [hunk 30]
```diff
@@ -51,6 +51,7 @@ export default ({ node }) => {
       commit(`setWalletAddress`, address)
       await dispatch(`queryWalletBalances`)
       dispatch(`loadDenoms`)
+      dispatch(`getTotalRewards`)
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/wallet.js]: you can instead use a `initializeWallet` action in the `distribution` module which will in parallel be triggered. this way the code is more modular and domain specific.

File: app/src/renderer/vuex/modules/distribution.js [hunk 31]
```diff
@@ -66,9 +66,6 @@ export default ({ node }) => {
         await dispatch(`getTotalRewards`)
       }
     },
-    async initializeWallet({ dispatch }) {
-      await dispatch(`getTotalRewards`)
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/distribution.js]: lol this was already here. let's do it this way, see other comment.

File: app/src/renderer/components/staking/LiValidator.vue (hunk 27 above)
Comment (faboweb) [reply app/src/renderer/components/staking/LiValidator.vue]: `calculateTokens` returns `BigNumber`. this way the return type stays predictable.

File: app/src/renderer/vuex/modules/wallet.js (hunk 30 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/wallet.js]: `initializeWallet` it's not distribution module specific

Comment (fedekunze) [reply app/src/renderer/vuex/modules/wallet.js]: I'd prefer to call everything on the original `initializeWallet` action

Comment (faboweb) [reply app/src/renderer/vuex/modules/wallet.js]: let's rename `initializeWallet` to `initializeAccount` or `initializeSession` then

File: app/src/renderer/components/staking/LiValidator.vue (hunk 27 above)
Comment (sabau) [reply app/src/renderer/components/staking/LiValidator.vue]: nice, got it thanks!

File: app/src/renderer/components/common/TmBalance.vue (hunk 29 above)
Comment (sabau) [reply app/src/renderer/components/common/TmBalance.vue]: yeah I missed this one, same as `PageNetwork`

File: test/unit/specs/store/distribution.spec.js (hunk 28 above)
Comment (fedekunze) [reply test/unit/specs/store/distribution.spec.js]: I like to test if the dispatches were called with the right parameters, happy to discuss it afterwards tho

File: app/src/renderer/components/network/PageNetwork.vue (hunk 26 above)
Comment (fedekunze) [reply app/src/renderer/components/network/PageNetwork.vue]: > can be moved to a function that takes as argument an integer

I like this idea, I'd add also a string parameter for the default response too (e.g: `n/a`). We could move this into a helper function in `scripts` since we use it in more pages as well

File: app/src/renderer/components/network/PageNetwork.vue [hunk 32]
```diff
@@ -40,23 +40,19 @@
             <dl class="info_dl">
               <dt>Total Liquid {{ bondDenom }}</dt>
               <dd id="loose_tokens">
-                {{ pool.pool.not_bonded_tokens ? num.pretty(pool.pool.not_bonded_tokens) : `n/a` }}
+                {{ pool.pool.not_bonded_tokens ? num.pretty(
```
Comment (faboweb) [inline app/src/renderer/components/network/PageNetwork.vue]: defensive programming
